### PR TITLE
SONARAZDO-384 Make shelljs dependency usable in our task codes

### PR DESCRIPTION
**Pull request by 7PH**
[ticket](https://sonarsource.atlassian.net/browse/SONARAZDO-384)

This will unblock [SONARAZDO-379](https://sonarsource.atlassian.net/browse/SONARAZDO-379)

The second commit fixes a regression I introduced [in this PR](https://github.com/SonarSource/sonar-scanner-azdo/pull/409) where the QA was trying to run before the extension was installed and tasks are available

[SONARAZDO-379]: https://sonarsource.atlassian.net/browse/SONARAZDO-379?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -159,11 +153,26 @@ stages:
               publisherId: "$(publisher)"
               extensionId: "sonarcloud"
               accounts: "https://devops.azure.com/$(itOrg)"
+          - task: Bash@3
+            displayName: "Wait for tasks to be available in CI"
+            env:
+              SERVICE_URL: "https://devops.azure.com/$(AzdoItsOrga)"
+              AZURE_TOKEN: $(ItOrgPat)
```
Comment (7PH) [inline azure-pipelines.yml:160]: I created a new PAT `ItOrgPat` on AzDO variable group to be able to run `tfx` in the test org

File: gulpfile.js [hunk 2]
```diff
@@ -191,10 +193,14 @@ gulp.task("build:copy", () => {
      * Hotfix for shelljs dependency
      * @see https://github.com/microsoft/azure-pipelines-task-lib/issues/942#issuecomment-1904939900
      */
-    const createShellJsDummyFile = gulpFile("index.js", "", { src: true }).pipe(
-      gulp.dest(path.join(outPath, "node_modules", "shelljs")),
-    );
-    streams.push(createShellJsDummyFile);
+    // eslint-disable-next-line import/no-dynamic-require
+    const esbuildConfig = require(path.join(SOURCE_DIR, "common", commonPath, "esbuild.config.js"));
+    if (esbuildConfig.external?.includes("shelljs")) {
```
Comment (7PH) [inline gulpfile.js:198]: We stop mocking `shelljs` library when `shelljs` isn't specified in externals

File: azure-pipelines.yml (hunk 1 above)
Comment (lucas-paulger-sonarsource) [reply azure-pipelines.yml:160]: seems I need this change in my PR too as the race condition is present on master, will wait for this to be merged to rebase

File: azure-pipelines.yml [hunk 3]
```diff
@@ -159,11 +153,26 @@ stages:
...
+            inputs:
+              targetType: "inline"
+              script: |
+                extension_installed=""
+                while [ -z "$extension_installed" ]; do
+                  echo "Waiting for SonarCloudPublish task to be registered..."
+                  tasks=$(tfx build tasks list --authType pat --service-url "$SERVICE_URL" -t "$AZURE_TOKEN" --no-color --json)
```
Comment (lucas-paulger-sonarsource) [inline azure-pipelines.yml:167]: nice simple iterative approach to grab the extension id when it's available!

File: gulpfile.js [hunk 4]
```diff
@@ -123,6 +124,7 @@ gulp.task("build:bundle", async () => {
       ...esbuildConfig,
       entryPoints: [task],
       outfile: outFilePath,
+      absWorkingDir: path.join(SOURCE_DIR, "common", commonFolder),
```
Comment (lucas-paulger-sonarsource) [inline gulpfile.js:127]: is `absWorkingDir` an esbuild standard field? I can't find it specifically mentioned in the API
https://esbuild.github.io/api/

but I see other plugins use it

Comment (7PH) [reply gulpfile.js:127]: image

yee you can see it in the JS API

File: src/common/latest/esbuild.config.js [hunk 5]
```diff
@@ -1,11 +1,10 @@
+const { shellJsPlugin } = require("esbuild-plugin-shelljs");
```
Comment (lucas-paulger-sonarsource) [inline src/common/latest/esbuild.config.js:1]: looking into your plugin, TIL what `npx patch-package` is 🎓 

very neat fix!

**Review by lucas-paulger-sonarsource — APPROVED** [review]
LGTM

File: gulpfile.js (hunk 4 above)
Comment (lucas-paulger-sonarsource) [reply gulpfile.js:127]: oof that was well hidden 🙏🏻 

https://esbuild.github.io/api/#working-directory (JS tab)

**Pull request merged**